### PR TITLE
Log cleanup for program operation in ProgramLifecycleHTTPHandler

### PR DESCRIPTION
**Pull request by rohitsinha54**
Some more log cleanup in ProgramLifecycleHttpHandler

Build: http://builds.cask.co/browse/CDAP-DUT3000-1

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package co.cask.cdap.common;
+
+/**
+ * Thrown when some method is not allowed.
+ */
+public class MethodNotAllowed extends Exception {
+
+  public MethodNotAllowed(String message) {
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: should probably store the method that is not allowed, and then construct the message, for consistency?

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.common;
+
+/**
+ * Thrown when some method is not allowed.
+ */
+public class MethodNotAllowed extends Exception {
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: ``MethodNotAllowedException`` for consistency?

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: 2015

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java [hunk 4]
```diff
@@ -16,6 +16,9 @@
 
 package co.cask.cdap.test.internal;
 
+import co.cask.cdap.common.BadRequestException;
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java]: unused?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 5]
```diff
@@ -422,11 +423,11 @@ public void programHistory(HttpRequest request, HttpResponder responder,
...
                              @QueryParam("end") String endTs,
-                             @QueryParam("limit") @DefaultValue("100") final int resultLimit) {
-    ProgramType type = ProgramType.valueOfCategoryName(programType);
+                             @QueryParam("limit") @DefaultValue("100") final int resultLimit) throws
+                                                                                BadRequestException, MethodNotAllowed {
+    ProgramType type = getProgramType(programType);
     if (type == null || type == ProgramType.WEBAPP) {
-      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
-      return;
+      throw new MethodNotAllowed(String.format("Program history is not supported for program type '%s'.", programType));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: backward-incompatible change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 6]
```diff
@@ -443,24 +444,21 @@ public void programRunRecord(HttpRequest request, HttpResponder responder,
                              @PathParam("app-id") String appId,
                              @PathParam("program-type") String programType,
                              @PathParam("program-id") String programId,
-                             @PathParam("run-id") String runid) {
-    ProgramType type = ProgramType.valueOfCategoryName(programType);
+                             @PathParam("run-id") String runid) throws NotFoundException, MethodNotAllowed {
+    ProgramType type = getProgramType(programType);
     if (type == null || type == ProgramType.WEBAPP) {
-      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+      throw new MethodNotAllowed(String.format("Program run record is not supported for program " +
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: ditto

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 7]
```diff
@@ -472,19 +470,19 @@ public void getProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
...
                                     @PathParam("program-type") String programType,
-                                    @PathParam("program-id") String programId) {
-    ProgramType type = ProgramType.valueOfCategoryName(programType);
+                                    @PathParam("program-id") String programId) throws BadRequestException,
+    NotImplementedException, NotFoundException, MethodNotAllowed {
+    ProgramType type = getProgramType(programType);
     if (type == null || type == ProgramType.WEBAPP) {
-      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
-      return;
+      throw new MethodNotAllowed(String.format("Get program runtime arguments is not supported for program " +
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: ditto

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 8]
```diff
@@ -499,19 +497,20 @@ public void saveProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
...
                                      @PathParam("program-type") String programType,
-                                     @PathParam("program-id") String programId) {
-    ProgramType type = ProgramType.valueOfCategoryName(programType);
+                                     @PathParam("program-id") String programId) throws BadRequestException,
+                                                          NotImplementedException, NotFoundException, MethodNotAllowed {
+    ProgramType type = getProgramType(programType);
     if (type == null || type == ProgramType.WEBAPP) {
-      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
-      return;
+      throw new MethodNotAllowed(String.format("Save program runtime arguments is not supported for program " +
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: ditto

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 9]
```diff
@@ -522,25 +521,20 @@ public void saveProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
...
-      } else {
-        responder.sendJson(HttpResponseStatus.OK, specification);
-      }
-    } catch (SecurityException e) {
-      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    Id.Program id = Id.Program.from(namespaceId, appId, type, programId);
+    ProgramSpecification specification = getProgramSpecification(id);
+    if (specification == null) {
+      throw new NotFoundException(String.format("Specification not found for program %s", programId));
+    } else {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: no need for else

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java (hunk 1 above)
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: In fact, how about 
```
MethodNotAllowedException(HttpMethod method, String apiPath) {
  super(String.format("Method %s is not allowed on path %s.", method, apiPath);
}
```

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java (hunk 2 above)
Comment (anwar6953) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: I was considering suggesting the same.
:+1:

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java (hunk 1 above)
Comment (rohitsinha54) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowed.java]: Nice :+1:

**Comment by rohitsinha54** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT3000-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 10]
```diff
@@ -438,11 +439,12 @@ public void programHistory(HttpRequest request, HttpResponder responder,
                              @QueryParam("status") String status,
                              @QueryParam("start") String startTs,
                              @QueryParam("end") String endTs,
-                             @QueryParam("limit") @DefaultValue("100") final int resultLimit) {
-    ProgramType type = ProgramType.valueOfCategoryName(programType);
+                             @QueryParam("limit") @DefaultValue("100") final int resultLimit) throws
+    BadRequestException, NotFoundException {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: nit: bring the 'throws' to the next line for better alignment

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 11]
```diff
@@ -459,24 +461,21 @@ public void programRunRecord(HttpRequest request, HttpResponder responder,
...
-    } catch (SecurityException e) {
-      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    Id.Program progId = Id.Program.from(namespaceId, appId, type, programId);
+    RunRecordMeta runRecordMeta = store.getRun(progId, runid);
+    if (runRecordMeta != null) {
+      RunRecord runRecord = CONVERT_TO_RUN_RECORD.apply(runRecordMeta);
+      responder.sendJson(HttpResponseStatus.OK, runRecord);
+      return;
     }
+    throw new NotFoundException(String.format("No run record information found for program %s with run id %s", progId,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: You should just be able to ``throw new NotFoundException(Id.Run)`` here, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 12]
```diff
@@ -538,26 +538,19 @@ public void saveProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
...
-        responder.sendStatus(HttpResponseStatus.NOT_FOUND);
-      } else {
-        responder.sendJson(HttpResponseStatus.OK, specification);
-      }
-    } catch (SecurityException e) {
-      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
+    Id.Program id = Id.Program.from(namespaceId, appId, type, programId);
+    ProgramSpecification specification = getProgramSpecification(id);
+    if (specification == null) {
+      throw new NotFoundException(String.format("Specification not found for program %s", programId));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Just NotFoundException(programId)?

File: cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowedException.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+package co.cask.cdap.common;
+
+import org.jboss.netty.handler.codec.http.HttpMethod;
+
+/**
+ * Thrown when some method is not allowed.
+ */
+public class MethodNotAllowedException extends Exception {
+
+  private HttpMethod method;
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/MethodNotAllowedException.java]: would be good to add getters for these. Also, do we need the first constructor on line 29? If not, lets remove it and make these members final? We can add it if required later. But ideally, there shouldn't be a MethodNotALlowedException without a method and an apiPath.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java [hunk 14]
```diff
@@ -161,6 +161,10 @@ public boolean isRunning(Id.Program programId) {
 
   @Override
   public List<RunRecord> getHistory(Id.Program programId, ProgramRunStatus status) {
-    return appFabricClient.getHistory(programId, status);
+    try {
+      return appFabricClient.getHistory(programId, status);
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java]: hmm not possible to propagate a checked exception here? Or is this a pattern in this class?

**Comment by bdmogal** [comment]
@rohitsinha54 , a few more minor comments.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java (hunk 14 above)
Comment (rohitsinha54) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java]: Its a pattern in this class.

**Comment by rohitsinha54** [comment]
@bdmogal : Addressed comments. Please have a look. 

New Build: http://builds.cask.co/browse/CDAP-DUT3000-4

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 15]
```diff
@@ -1491,20 +1484,16 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
...
-        ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
-          ProgramRunStatus.valueOf(status.toUpperCase());
-        List<RunRecord> records =
-          Lists.transform(store.getRuns(programId, runStatus, start, end, limit), CONVERT_TO_RUN_RECORD);
-        responder.sendJson(HttpResponseStatus.OK, records);
-      } catch (IllegalArgumentException e) {
-        responder.sendString(HttpResponseStatus.BAD_REQUEST,
-                             "Supported options for status of runs are running/completed/failed");
-      }
-    } catch (SecurityException e) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: why was this removed?

**Comment by bdmogal** [comment]
one more question. rest LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 15 above)
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: We don't need to catch SecurityExcepetion here and handle it the HTTPExceptionHandler does it. We are removing it from all the methods in this class.
